test/support/utils_test_base.h

```diff
@@ -0,0 +1,717 @@
+        ::std::size_t    offset = 0;        // Offset in USM buffer
+
+        template<typename _Size>
+        Data(sycl::usm::alloc __alloc_type, sycl::queue __q, _Size __sz, ::std::size_t __offset)
+            : alloc_type(__alloc_type)
+            , offset(__offset)
+        {
+            // We use this switch/case because we have test_base_data_visitor interface with
+            // virtual functions, arguments of which can't be template classes.
+            // So, USM allocation type specified in runtime.
+            switch (__alloc_type)
+            {
+            case sycl::usm::alloc::shared:
+                src_data_usm.reset(new usm_data_transfer<sycl::usm::alloc::shared, TestValueType>(__q, __sz + __offset));
+                break;
+            case sycl::usm::alloc::device:
+                src_data_usm.reset(new usm_data_transfer<sycl::usm::alloc::device, TestValueType>(__q, __sz + __offset));
+                break;
+            default:
+                assert(false);
+                break;
+            }
+        }
+
+        auto get_usm_data_shared()
+        {
+            assert(alloc_type == sycl::usm::alloc::shared);
+            return reinterpret_cast<usm_data_transfer<sycl::usm::alloc::shared, TestValueType>*>(src_data_usm.get());
+        }
+
+        auto get_usm_data_device()
+        {
+            assert(alloc_type == sycl::usm::alloc::device);
+            return reinterpret_cast<usm_data_transfer<sycl::usm::alloc::device, TestValueType>*>(src_data_usm.get());
+        }
+
+        TestValueType* get_start_from()
+        {
+            TestValueType* result = nullptr;
+
+            switch (alloc_type)
+            {
+            case sycl::usm::alloc::shared:
+                result = get_usm_data_shared()->get_data() + offset;
+                break;
+
+            case sycl::usm::alloc::device:
+                result = get_usm_data_device()->get_data() + offset;
+                break;
+
+            default:
+                assert(false);
+            }
+
+            return result;
+        }
+
+        template<typename _Iterator, typename TDiff>
+        void retrieve_data(_Iterator __it, TDiff __objects_count)
+        {
+            switch (alloc_type)
+            {
+            case sycl::usm::alloc::shared:
+                get_usm_data_shared()->retrieve_data(__it, offset, __objects_count);
+                break;
+
+            case sycl::usm::alloc::device:
+                get_usm_data_device()->retrieve_data(__it, offset, __objects_count);
+                break;
+
+            default:
+                assert(false);
+            }
+        }
+
+        template<typename _Iterator, typename TDiff>
+        void update_data(_Iterator __it, TDiff __objects_count)
+        {
+            switch (alloc_type)
+            {
+            case sycl::usm::alloc::shared:
+                get_usm_data_shared()->update_data(__it, offset, __objects_count);
```

Comment:
the same comment about get_usm_data_shared

Reply:
Fixed: added comments